File: server/src/main/java/logState/LogState.java
package logState;

import battle.BattleMgr;
import config.battle.MapConfig;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;

public class LogState {

    private static LogState single_instance = null;

    public static LogState getInstance() {
        if (single_instance == null) {
            single_instance = new LogState();
        }
        return single_instance;
    }

    public void logToFile(BattleMgr gameMgr, int countLoop){
        String content = "";

        // Log monster
        content += "Player: [x - y - numBomb - powerBlast - numberOfLife - speed]";
        double sumPlayer = 0;
        for (int i = 0; i < gameMgr.listPlayer.size(); i++) {
            content += "\n" + i + ". ["
                    + String.format("%.3f", gameMgr.listPlayer.get(i).currentPos.getX()) + " - "
                    + String.format("%.3f", gameMgr.listPlayer.get(i).currentPos.getY()) + "-"
                    + gameMgr.listPlayer.get(i).numbMaxBomb + "-"
                    + gameMgr.listPlayer.get(i).powerBlast + "-"
                    + gameMgr.listPlayer.get(i).numberOfLife + "-"
                    + String.format("%.3f", gameMgr.listPlayer.get(i).speed) + "]";

            sumPlayer += gameMgr.listPlayer.get(i).currentPos.getX()
                    + gameMgr.listPlayer.get(i).currentPos.getY()
                    + gameMgr.listPlayer.get(i).numbMaxBomb
                    + gameMgr.listPlayer.get(i).powerBlast
                    + gameMgr.listPlayer.get(i).numberOfLife
                    + gameMgr.listPlayer.get(i).speed;
        }

        // Log list bomb
        content += "\n\nBomb: [userID - x - y - timer]";
        double sumBomb = 0;
        for (int i = 0; i < gameMgr.listBomb.size(); i++) {
            content += "\n" + i + ". ["
                    + gameMgr.listBomb.get(i).player.userId + " - "
                    + String.format("%.3f", gameMgr.listBomb.get(i).pos.getX()) + " - "
                    + String.format("%.3f", gameMgr.listBomb.get(i).pos.getY()) + " - "
                    + String.format("%.3f", gameMgr.listBomb.get(i).timer) + "]";

            sumBomb += gameMgr.listBomb.get(i).player.userId
                    + gameMgr.listBomb.get(i).pos.getX()
                    + gameMgr.listBomb.get(i).pos.getY()
                    + gameMgr.listBomb.get(i).timer;
        }

        // Log list bullet
        content += "\n\nBullet: [userID - x - y - isDestroy]";
        double sumBullet = 0;
        for (int i = 0; i < gameMgr.listBullet.size(); i++) {
            content += "\n" + i + ". ["
                    + gameMgr.listBullet.get(i).player.userId + " - "
                    + String.format("%.3f", gameMgr.listBullet.get(i).currentPos.getX()) + " - "
                    + String.format("%.3f", gameMgr.listBullet.get(i).currentPos.getY()) + " - "
                    + gameMgr.listBullet.get(i).isDestroy + "]";

            sumBullet += gameMgr.listBullet.get(i).player.userId
                    + gameMgr.listBullet.get(i).currentPos.getX()
                    + gameMgr.listBullet.get(i).currentPos.getY();
        }

        // Log map state
        content += "\n\nMap logic:";
        double sumMap = 0;
        for (int i = 0 ; i < gameMgr.mapMgr.size; i++) {
            content += "\n[";
            for (int j = 0; j < gameMgr.mapMgr.size; j++) {
                sumMap += gameMgr.mapMgr.data[i][j];
                content += gameMgr.mapMgr.data[i][j];
                if ( j != gameMgr.mapMgr.size - 1) {
                    content += " - ";
                }
            }
            content += "]";
        }

        // Round
        String contentSum = String.format("%.3f", sumPlayer) + "\n"
                            + String.format("%.3f", sumBomb) + "\n"
                            + String.format("%.3f", sumBullet) + "\n"
                            + sumMap;

        logState(content, countLoop);
        logSumState(contentSum, countLoop);
    }

    public void logState(String content, int countLoop) {
        try {
            // Make new directory
            String dir = "../logServer/logState";
            File theDir = new File(dir);
            if (!theDir.exists()){
                theDir.mkdirs();
            }

            // Make new file
            String path = dir + "/"+ countLoop + "-server-stateFrame.txt";
            File file = new File(path);
            file.createNewFile();

            // Write file
            FileWriter fw = new FileWriter(path, false);// ghi de
            BufferedWriter bw = new BufferedWriter(fw);
            bw.write(content);
            bw.close();

        } catch (Exception e) {
            System.out.println(e);
        }
    }

    public void logSumState(String content, int countLoop) {
        try {
            // Make new directory
            String dir = "../logServer/logSumState";
            File theDir = new File(dir);
            if (!theDir.exists()){
                theDir.mkdirs();
            }

            // Make new file
            String path = dir + "/" + countLoop + "-server-sumStateFrame.txt";
            File file = new File(path);
            file.createNewFile();

            // Write file
            FileWriter fw = new FileWriter(path, false);// ghi de
            BufferedWriter bw = new BufferedWriter(fw);
            bw.write(content);
            bw.close();

        } catch (Exception e) {
            System.out.println(e);
        }
    }
}

File: client/src/modules/LogState/LogState.js
var LogState = cc.Class.extend({

    ctor: function () {
        this.flag = false;
        this.dirState = "../logClient/logState/user" + gv.user.id + "/"
        this.dirSumState = "../logClient/logSumState/user" + gv.user.id + "/"
        jsb.fileUtils.createDirectory(this.dirState)
        jsb.fileUtils.createDirectory(this.dirSumState)

        this.deltaPlayer = 0.1;
        this.deltaBomb = 0.1;
        this.deltaBullet = 0.1;
        this.deltaMap = 0.1;
    },

    logToFile: function (gameMgr, countLoop) {
        let content = ""

        // Log player's position
        content += "\nPlayer: [x - y - numBomb - powerBlast - numberOfLife - speed]"
        let sumPlayer = 0
        for (let i = 0; i < gameMgr.listPlayer.length; i++) {
            content += "\n" + i + ". ["
                + gameMgr.listPlayer[i].currentPos.x.toFixed(3) + " - "
                + gameMgr.listPlayer[i].currentPos.y.toFixed(3) + "-"
                + gameMgr.listPlayer[i].numbMaxBomb + "-"
                + gameMgr.listPlayer[i].powerBlast + "-"
                + gameMgr.listPlayer[i].numberOfLife + "-"
                + gameMgr.listPlayer[i].speed.toFixed(3) + "]"

            sumPlayer += gameMgr.listPlayer[i].currentPos.x
                + gameMgr.listPlayer[i].currentPos.y
                + gameMgr.listPlayer[i].numbMaxBomb
                + gameMgr.listPlayer[i].powerBlast
                + gameMgr.listPlayer[i].numberOfLife
                + gameMgr.listPlayer[i].speed
        }

        // Log list bomb
        content += "\n\nBomb: [userID - x - y - timer]";
        let sumBomb = 0;
        for (let i = 0; i < gameMgr.listBomb.length; i++) {
            content += "\n" + i + ". ["
                + gameMgr.listBomb[i].player.userId + " - "
                + gameMgr.listBomb[i].pos.x.toFixed(3) + " - "
                + gameMgr.listBomb[i].pos.y.toFixed(3) + " - "
                + gameMgr.listBomb[i].timer.toFixed(3) + "]";

            sumBomb += gameMgr.listBomb[i].player.userId
                + gameMgr.listBomb[i].pos.x
                + gameMgr.listBomb[i].pos.y
                + gameMgr.listBomb[i].timer;
        }

        // Log list bullet
        content += "\n\nBullet: [userID - x - y - isDestroy]";
        let sumBullet = 0;
        for (let i = 0; i < gameMgr.listBullet.length; i++) {
            content += "\n" + i + ". ["
                + gameMgr.listBullet[i].player.userId + " - "
                + gameMgr.listBullet[i].currentPos.x.toFixed(3) + " - "
                + gameMgr.listBullet[i].currentPos.y.toFixed(3) + " - "
                + gameMgr.listBullet[i].isDestroy + "]";

            sumBullet += gameMgr.listBullet[i].player.userId
                + gameMgr.listBullet[i].currentPos.x
                + gameMgr.listBullet[i].currentPos.y;
        }

        // Log map state
        content += "\n\nMap logic:";
        let sumMap = 0;
        for (let i = 0 ; i < gameMgr.mapMgr.size; i++) {
            content += "\n[";
            for (let j = 0; j < gameMgr.mapMgr.size; j++) {
                sumMap += gameMgr.mapMgr.data[i][j];
                content += gameMgr.mapMgr.data[i][j];
                if ( j !== gameMgr.mapMgr.size - 1) {
                    content += " - ";
                }
            }
            content += "]";
        }

        // Get path
        let pathState = this.dirState + "/"+ countLoop +"-client-stateFrame.txt"
        let pathSumState = this.dirSumState + "/"+ countLoop +"-client-sumStateFrame.txt"

        this.dirSumStateServer = "../logServer/logSumState/"
        let pathSumStateServer = this.dirSumStateServer + "/"+ countLoop +"-server-sumStateFrame.txt"

        // Compare with server state
        let sumStateServer = jsb.fileUtils.getStringFromFile(pathSumStateServer)
        let sum = sumStateServer.split("\n");
        let svSumPlayer = Number(sum[0])
        let svSumBomb = Number(sum[1])
        let svSumBullet = Number(sum[2])
        let svSumMap = Number(sum[3])

        this.logError(sumPlayer, svSumPlayer, countLoop, "Player", this.deltaPlayer)
        this.logError(sumBomb, svSumBomb, countLoop, "Bomb", this.deltaBomb)
        this.logError(sumBullet, svSumBullet, countLoop, "Bullet", this.deltaBullet)
        this.logError(sumMap, svSumMap, countLoop, "Map", this.deltaMap)

        // Round
        sumPlayer = Utils.round(sumPlayer).toFixed(3);
        sumBomb = Utils.round(sumBomb).toFixed(3);
        sumBullet = Utils.round(sumBullet).toFixed(3);
        sumMap = Utils.round(sumMap).toFixed(3);
        let contentSum = sumPlayer + "\n" + sumBomb + "\n" + sumBullet + "\n" + sumMap;

        // Log state to file
        jsb.fileUtils.writeStringToFile(content, pathState);
        jsb.fileUtils.writeStringToFile(contentSum, pathSumState);
    },

    logError: function (client, server, countLoop, txt, delta) {
        if (Math.abs(client - server) > delta) {
            cc.log("\n" + txt + " - USER " + gv.user.id + " - ERROR STATE AT FRAME: " + countLoop + ", " + Math.abs(client - server) + "\n")

            switch (txt) {
                case "Player": {
                    this.deltaPlayer = delta + Math.abs(client - server)
                    break
                }
                case "Bomb": {
                    this.deltaBomb = delta + Math.abs(client - server)
                    break
                }
                case "Bullet": {
                    this.deltaBullet = delta + Math.abs(client - server)
                    break
                }
                case "Map": {
                    this.deltaMap = delta + Math.abs(client - server)
                    break
                }
            }
        }
    }
})

var _logState;
LogState.getInstance = function () {
    if (_logState === undefined)
        _logState = new LogState();
    return _logState;
}